api/v1/standalone-dns-proxy/standalone-dns-proxy.proto

```diff
@@ -0,0 +1,85 @@
+
+package standalonednsproxy;
+
+option go_package = "github.com/cilium/cilium/api/v1/standalonednsproxy";
+
+// Cilium agent runs the FQDNData service and Standalone DNS proxy connects to it to get the DNS Policy rules.
+// Standalone DNS proxy sends FQDN-IP mapping updates to Cilium Agent.
+// CFP: https://github.com/cilium/design-cfps/pull/54
+service FQDNData {
+  // StreamPolicyState is used by the Standalone DNS proxy to get the current policy state. 
+  // Policy state includes the DNS policies and the identity to IP mapping.
+  // Cilium agent will stream DNS policies state to Standalone DNS proxy.
+  // In case of any client side error, cilium agent will cancel the stream and SDP will have to re-subscribe.
+  // In case of any server side error, cilium agent will send an error response and SDP will have to re-subscribe.  
+  rpc StreamPolicyState(stream PolicyStateResponse) returns (stream PolicyState) {}
+  
+  // UpdateMappingRequest is used by the Standalone DNS proxy to update ciliium agent with FQDN-IP mappings which in turn update L3/L4 policy maps.
+  // In case of any error, SDP will either retry the connection if the error is server side or will error out.
+  // Note: In case of concurrent updates, since this is called in a callback(notifyDNSMsg) from the DNS server it follows the same behavior as
+  // the inbuilt dns proxy in cilium.
+  rpc UpdateMappingRequest(FQDNMapping) returns (UpdateMappingResponse) {}
+}
+
+// Response code returned by RPC methods.
+enum ResponseCode {
+    RESPONSE_CODE_UNSPECIFIED = 0;
+    RESPONSE_CODE_NO_ERROR = 1;
+    RESPONSE_CODE_FORMAT_ERROR = 2;
+    RESPONSE_CODE_SERVER_FAILURE = 3;
+    RESPONSE_CODE_NOT_IMPLEMENTED = 4;
+    RESPONSE_CODE_REFUSED = 5;
+}
+
+// Ack sent from SDP to Agent on processing DNS policy rules
+message PolicyStateResponse  {
+    ResponseCode response = 1;
+    string request_id = 2; // Request ID for which response is sent to
+}
+
+// FQDN-IP mapping goalstate sent from SDP to agent 
+message FQDNMapping {
+    string fqdn = 1; // dns name
+    repeated bytes record_ip = 2; // List of IPs corresponding to dns name
+    uint32 ttl = 3;  // TTL of DNS record
+    uint32 source_identity = 4; // Identity of the client making the DNS request
+    bytes source_ip = 5; // IP address of the client making the DNS request
+    uint32 response_code = 6; // DNS Response code as specified in RFC2316
+}
+
+// Ack returned by cilium agent to SDP on receiving FQDN-IP mapping update
+message UpdateMappingResponse {
+    ResponseCode response = 1;
+}
+
+// DNServer identity, port and protocol the requests be allowed to
+message DNSServer {
+    uint32 dns_server_identity = 1;  // Identity of destination DNS server
+    uint32 dns_server_port = 2;
+    uint32 dns_server_proto = 3;
+}
+
+// L7 DNS policy specifying which requests are permitted to which DNS server
+message DNSPolicy {
+    uint32 source_identity = 1;  // Identity of the workload this L7 DNS policy should apply to
```

Comment:
I was under the impression that the dns proxy would apply egress policy by endpoint id not sec id, is this really the ID you need?

Reply:
SDP enforces egress policy based on identity and not on endpoint ID.. When SDP receives DNS request , it will lookup for source ip->identity mapping and then get rules associated with that identity and enforce it.. In case of in-agent dns proxy it works based on endpoint ID which we can update similar to SDP in future. If it's based on endpoint ID, then agent have to send policy for each endpoint ID which will be a duplication and since identity can map to multiple endpointID, it makes sense for us to enforce based on identity instead of endpoint ID.

Reply:
Why? The existing DNSproxy already has logic to enforce the egress DNS policy, why duplicate it? 
I've now looked, and in the full PR you've reimplemented policy based on the security identity. To me it's surprising, and probably not a good idea to have two implementations of the same thing?

IMO either we have something which maps IPs to endpoint IDs in the API or we change the implementation in the agent to be based on security identity too.

(for context, I'm talking about 
https://github.com/cilium/cilium/pull/36214/files#diff-a7f128716bd349970eb11e4decfff9d79cb9daa1067020b9c9a7efd664bd8034R1126-R1173 here which is a bit confusing, but I guess this has an effect on the gRPC API 🤷‍♂️ )

Reply:
I hear your point. That was our initial proposal to keep logic same as agent dnsproxy. As discussion went ahead in cfp, we designed in a way that what's the minimum data required by sdp to solve this and found it just requires ip and identity. endpoint id is kind of duplicate and ignored sending it.

> IMO either we have something which maps IPs to endpoint IDs in the API or we change the implementation in the agent to be based on security identity too.

I agree on this. Looking at dnsproxy code, it's not straight forward (especially restoration logic) to change its logic to enforce based on identity. we will stick to enforce policy based on endpoint id instead of identity as like agent dnsproxy.

Reply:
That's fine by me. Personally also don't particularly like this dependency on EPID, but maybe that's a topic for another PR. Fine as is.